### DISCUSSION TITLE
How to use the ReducerProtocol approach for a feature whose state contains an optional enum?

### DISCUSSION THREAD

**Discussion by ghost** [pointfreeco/swift-composable-architecture#1615] [Q&A] [answered]
Hello all,

In our app, we have modelled navigation by having a `Route` enumerated type (with states for associated types) and a matching `NavigationRoute` enumerated type. The `NavigationRoute` type exists simply to extract the name of a `Route` case, without caring for the associated state, and works as a `tag` when using the navigation APIs. We do this to avoid having to require `State`s to be `Hashable` since the `tag`s must be.

```swift
extension FooFeature {
   struct State: Equatable {
      var route: Route?
      var navRoute: NavigationRoute? { route?.navRoute }
      // ...
   }
}

extension FooFeature.State {

   enum Route: Equatable {
      case chat(ChatFeature.State)
      case participants(ParticipantsFeature.State)
      // ...

      var navRoute: NavigationRoute {
         switch self {
            case .chat(_):
               return .chat
            case .participants(_):
               return .participants
            // ...
         }
      }
   }

   enum NavigationRoute: Hashable {
      case chat
      case participants
      // ...
   }

}

extension FooFeature {
   enum Action: Equatable {
      case chat(ChatFeature.Action)
      case participants(ParticipantsFeature.Action)
      // ...
   }
}

struct FooFeature: ReducerProtocol {
   var body: some ReducerProtocol<State, Action> {
      Reducer { state, action in
         switch action {
            // ...
         }
      }
     // what goes in here ?
   }
}
```

It seems to me that this is a combination of the "optional state" and "enum state" ways of modelling state so I thought some combination of the `.ifLet()` and `.ifCaseLet()` reducer modifiers would be called for but I couldn't make that work.

Any help would be greatly appreciated. Thanks!

Addendum:

Incidentally, before the `ReducerProtocol` approach, we created a custom pullback, like so:
```swift
public extension Reducer {
    func pullback<GlobalState, GlobalAction, GlobalEnvironment, Route>(
        unwrapping toRoute: WritableKeyPath<GlobalState, Route?>,
        case toRouteCase: CasePath<Route, State>,
        action toLocalAction: CasePath<GlobalAction, Action>,
        environment toLocalEnvironment: @escaping (GlobalEnvironment) -> Environment
    ) -> Reducer<GlobalState, GlobalAction, GlobalEnvironment> {
        .init { globalState, globalAction, globalEnvironment in

            // do we have the correct child action?
            guard
                let localAction = toLocalAction.extract(from: globalAction)
            else { return .none }

            guard
                // is a child actually being presented?
                let route = globalState[keyPath: toRoute],
                // if so, do we have the right child/reducer pair?
                let routeCase = toRouteCase.extract(from: route)
            else { return .none } // if not, don't run the child reducer

            // get the child state
            var localState = routeCase

            // make sure to pass the (possibly updated) child state back to the parent
            defer { globalState[keyPath: toRoute] = toRouteCase.embed(localState) }

            // run the child reducer
            let effects = self.run(
                &localState,
                localAction,
                toLocalEnvironment(globalEnvironment)
            )
            .map(toLocalAction.embed)

            return effects
        }
    }
}
```

**Comment by tgrapperon** [comment, marked as answer]
Hey @wtruppel. Because of the "modifier" style approach that was selected, ~there is no "convenient" way to scope into optional enums right now~ (wrong, see @stephencelis answer below). One solution is to open the optional, and then open the enum's case using an `EmptyReducer` as a support:
```swift
Reduce { state, action in
  switch action {
    …  
  }
}
.ifLet(\.route, action: /Action.route) {            // Route? -> Route
  EmptyReducer()
    .ifCaseLet(/Route.chat, action: /Action.chat) { // Route -> Route.case
      ChatFeature()
    }
    .ifCaseLet(/Route.participants, action: /Action.participants) {
      ParticipantsFeature()
    }
}
```
Once this is working, you can try to make it work with your `NavigationRoute` enum, which I think is accessory to the problem. There is no ordering problem here because the initial `ifLet` runs all these reducers before the `Reduce` one.

Another approach that may help is to "coalesce" the nilness of `Route?` into `Route` itself by adding a `.none` case to `Route` [^1]. In this case, you wouldn't need the `ifLet(\.route, …)`.

[^1]: Fun fact, you can conform `Route` to `NilLiteralConvertible`, so you can still use `nil` to create `Route.none`. But it absolutely isn't worth the hassle!

**Reply by ghost** [reply]
Hi @tgrapperon, once again thank you for answering my questions!

> Another approach that may help is to "coalesce" the nilness of Route? into Route itself by adding a .none case to Route. 

I tried that but couldn't make it work (and it also means a large refactoring of our code).

I like your first suggestion, if I can make it work. Thanks again!

**Reply by ghost** [reply]
Oh, but wait... your first suggestion requires that I create an `Action` enum to go along with the `Route`.

**Reply by tgrapperon** [reply]
Ah sorry, you can directly use `.self` (the identity `CasePath`) that will propagate `Action` to the other side:
```swift
.ifLet(\.route, action: .self) { … } 
```

**Reply by ghost** [reply]
Ah.. cool. Thanks!

**Reply by ghost** [reply]
@tgrapperon 
The compiler didn't like `.ifLet(\.route, action: .self) { … } `. Even being more explicit, with `.ifLet(\.route, action: /Action.'self') { … }` didn't resolve the issue. The compiler "can't resolve the types in a reasonable time."

I don't understand how it was supposed to work. `Action` is a subtype of `FooFeature` and `Route` is a subtype of `State`, which is itself a subtype of `FooFeature`. How does `/Action.'self'` pass the right action through the layers of reducers being composed?

**Reply by tgrapperon** [reply]
~`/Action.self` won't work~. Your problem is most likely caused by the typo you have at `Reduce` where you used `Reducer` instead (and that I copy/pasted in my example):
```swift
Reduce[r] { state, action in
  switch action {
    …  
  }
}
```
If this doesn't help, you can try to be explicit with the `CasePath`: 
```swift
.`self` -> CasePath<Action, Action>.`self`
```
I may have forgotten the backticks around `.self` which are maybe required. I'm currently dealing with a water leak at home, so sorry if I wasn't super attentive.

**Reply by ghost** [reply]
No apologies necessary!
The typo was just here, not in the actual code. I'm having lunch so will try your latest suggestion afterwards. I hope your water leak gets resolved soon. :)

**Reply by tgrapperon** [reply]
> How does /Action.'self' pass the right action through the layers of reducers being composed?

It doesn't, it simply passes the `Action` as it, as it's directly opened on a case by case basis by `.ifCaseLet`. You may be familiar with mathematical analogies, so in this case, this is the identity (or trivial) transform of the (endo)morphism `Action -> Action` that you need here (as you don't want to open `Action` at this stage). If you had an `Action.route(Route.Action)` case, you would need to provide a `CasePath<Action, Route.Action>` which would be a morphism `Action` -> `Route.Action`.

**Reply by ghost** [reply]
Ah, my bad. It was a typo but not the one with the Reducer. In my actual code, the state and actions are called `participantsList`, not `participants` so when I tried your suggestion, I forgot to update them and the compiler choked. Anyway, your first suggestion works perfectly, including using `/Action.self` (no back-quotes needed but specifying the type, `Action`, was needed). One more time, many thanks for all your help.

**Reply by stephencelis** [reply]
Just to add to @tgrapperon's answer, you can use `Scope` to avoid the `EmptyReducer`-`ifCaseLet` dance:

```swift
Reduce { state, action in
  switch action {
    …  
  }
}
.ifLet(\.route, action: /Action.route) {
  Scope(state: /Route.chat, action: /Action.chat) {
    ChatFeature()
  }
  Scope(state: /Route.participants, action: /Action.participants) {
    ParticipantsFeature()
  }
}
```

While we prefer `ifCaseLet` when mixed with other logic that can switch the cases, the cases are mutually exclusive inside the `ifLet`, so it's safe to use `Scope` and save a layer of indentation and indirection instead.

**Reply by stephencelis** [reply]
(Though if you aren't pulling back through multiple layers of actions and are using the `/.self` case path, that might not be possible.)

**Reply by ghost** [reply]
Hi @stephencelis. Thanks for the extra info. We're not using another layer of actions so the `/.self` solution worked great.

**Reply by tgrapperon** [reply]
@stephencelis Indeed, for some reason, I forgot about `Scope`'s `CasePath`s overloads.
EDIT: And also the `/.self` and `/Action.self` overloads.
